che-incubator/che-code
# Send telemetry events to telemetry plugin

Signed-off-by: David Kwon <dakwon@redhat.com>

Part of https://github.com/eclipse/che/issues/21661

This PR introduces a new built-in extension:
![image](https://user-images.githubusercontent.com/83611742/192374112-052950e1-b374-4124-a3ae-6d9ff59fbe62.png)

Which sends telemetry events to `http://localhost:${DEVWORKSPACE_TELEMETRY_BACKEND_PORT}`

Compared to [Che-theia telemetry](https://github.com/eclipse/che/issues/15469), the Che-code telemetry:

* does not detect the `git.push` and `git.commit` commands ([like in che-theia](https://github.com/eclipse-che/che-theia/blob/8bef58f2223607312e09d7cca62d6df0c00f0ca6/plugins/telemetry-plugin/src/telemetry-plugin.ts#L17-L23)) since `onDidExecuteCommand` does not exist in the VS Code api: https://github.com/microsoft/vscode/issues/78091#issuecomment-518545491. Therefore the `Push to remote in Che`and `Commit locally in Che` telemetry events are not being detected/sent.
* does not detect screen dimensions since the `window` object is not available

### To test this PR:

1. Start a workspace with this URL
```
{CHE_HOST}/#https://github.com/dkwon17/che-code/tree/telemetryService
```

2. After the workspace starts, add the following telemetry plugin under the DevWorkspace's `spec.template.components`:
```
- attributes:
  name: telemetry
  plugin:
    uri: https://gist.githubusercontent.com/dkwon17/d2f800494959dadd1106512b0c56275e/raw/5fada3206ee6dba868fd73086acc21d46682cce8/woopra-test.yaml
```
This telemetry plugin will log the received events.
3. Start the workspace again
4. Run `yarn && yarn watch` and then run `yarn server` to install dependencies and start up a new instance of VS Code.
5. Click on `Open In New Tab` to open the new instance
![image](https://user-images.githubusercontent.com/83611742/193689155-1e6780af-4496-450f-9491-f4767e08690c.png)
6. View the container logs for the telemetry plugin. You could try to send and verify that the `Start Workspace in Che`, `Open Workspace in Che`, and `Edit Workspace File in Che` events are being sent:

`Start Workspace in Che` log (when opening the workspace)
```
2022-10-03 21:40:14,489 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Start Workspace in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723}
```

`Refresh Workspace in Che` log (you should see this when you refresh the page with F5 for example)
```
2022-10-03 21:51:09,529 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Refresh Workspace in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723}
```

`Edit Workspace File in Che` log (when editing a file)
```
2022-10-03 21:40:50,852 DEBUG [com.red.dev.ser.tel.woo.AnalyticsManager] (executor-thread-0) sending Edit Workspace File in Che (ip=null - userAgent=null) with properties: {creation time=1664830321000, devworkspace id=workspaceec5d20f7242b499b, devworkspace name=che-code, start time=1664833044918, age=2723, programming language=json}
```

There are more [telemetry events](https://github.com/che-incubator/che-workspace-telemetry-client/blob/4a6e6a8eae0be050a370823c6970faf182c5365e/backend-base/src/main/java/org/eclipse/che/incubator/workspace/telemetry/base/AnalyticsEvent.java#L17-L24), but the `Keep Workspace Inactive in Che` and `Stop Workspace in Che` events are handled form the plugin side and not the VS Code extension side (See [1](https://github.com/che-incubator/devworkspace-telemetry-woopra-plugin/blob/34bf91ed407ca49ceea1256c20599f72137ddee0/src/main/java/com/redhat/devworkspace/services/telemetry/woopra/AnalyticsManager.java#L188-L192), [2](https://github.com/che-incubator/devworkspace-telemetry-woopra-plugin/blob/34bf91ed407ca49ceea1256c20599f72137ddee0/src/main/java/com/redhat/devworkspace/services/telemetry/woopra/AnalyticsManager.java#L431-L432)).
